boards: added mikroelektronika esp32 wifi shield (#45548)
zephyrproject-rtos/zephyr: boards/shields/mikroe_wifi_bt_click/doc/index.rst

available at Espressif Systems web site. This version is command
compatible with ESP32 AT Bin 2.0.0, after getting the binary from Espressif
site, connect the J1 of the board to a serial-to-USB converter of your
preference, or to one Espressif programming boards like ESP-Prog. Install
the `ESP-Tool`_, extract the folder, navigate inside the binaries location
and type the following command to flash the device:

.. code-block:: console
   esptool.py --chip auto \
   --before default_reset \
   --after hard_reset write_flash \
   --flash_mode dio \
   --flash_freq 40m \
   --flash_size detect \
   0x10000 ota_data_initial.bin \
   0x1000 bootloader/bootloader.bin \
   0x20000 at_customize.bin \
   0x21000 customized_partitions/ble_data.bin \
   0x24000 customized_partitions/server_cert.bin \
   0x26000 customized_partitions/server_key.bin \
   0x28000 customized_partitions/server_ca.bin \
   0x2a000 customized_partitions/client_cert.bin \
   0x2c000 customized_partitions/client_key.bin \
   0x2e000 customized_partitions/client_ca.bin \
   0x30000 customized_partitions/factory_param.bin \
   0xf000 phy_init_data.bin \
   0x100000 esp-at.bin \

>    0x8000 partitions_at.bin

Reviewer: @mniestroj  Can we address this, at worst, in a documentation update to be able to get this in before the feature freeze today? The command above is valid, and is provided to be helpful, but we're free to improve this before the final 3.1 tag as a documentation fix.

>    0x8000 partitions_at.bin

Reviewer 2: Sure, that is not a blocker from my side.

.. _mikroe_wifi_bt_click:

MikroElektronika WIFI and BLE Shield
####################################

Overview
********

This shield contains an Espressif Systems ESP32 which is responsible
to offload the WIFI and Bluetooth tasks from the host processor. It uses
the standard MikroBus interface with UART pins plus its Espressif AT
command set. Also Zephyr modem driver could be used in place, to used
this module as a native WIFI networking interface

.. image:: ./wifi_click.jpeg
   :width: 634px
   :align: center
   :alt: ESP-01

> Pins Assignment of the ESP-01 Module

Reviewer 3: I think ESP-01 only works with CONFIG_WIFI_ESP_AT_VERSION_1_7 because of flash size, at least that is the last info that I have. Is espressif released a 2.0 version for ESP-01 that fits on ESP-01?

Author: Another copy pasta.

=============================

+-----------------------+---------------------------------------------+
| Shield Connector Pin  | Function                                    |
+=======================+=============================================+

> | RST#                  | ESP32 Reset                                 |

Reviewer: We should add a note that there is a conflict with the rst polarity and we need to disconnect the RST pin on most board (the LPC55S69 at least).

| Shield Connector Pin  | Function                                    |
+=======================+=============================================+
| RST#                  | Reset pin, in most of the boards it should  |
|                       | be disconnected because its tied to GND     |
+-----------------------+---------------------------------------------+
| TX                    | UART TX communication pin                   |
+-----------------------+---------------------------------------------+
| RX                    | UART RX communication pin                   |
+-----------------------+---------------------------------------------+

Requirements
************

This shield requires a board which provides a configuration that allows an
UART interface. (see :ref:`shields` for more details).

.. note::
   Sometimes boards declare standard headers like Arduino R3 but don't
   define all connections.  Make sure that the board you are using includes
   all required definitions to avoid build errors.

This shield should be loaded with the `ESP32 AT Bin`_ software which is
available at Espressif Systems web site. This version is command

> compatible with ESP32 AT Bin 2.0.0

Reviewer: Sorry to nag, but could you perhaps add a comment and code block here showing how people can flash the said firmware.

It takes a bit of digging to find the right command, and it's important to be able to reproduce the same device config used by this PR:

```bash
$ esptool.py \
  --chip auto \
  --baud 115200 \
  --before default_reset \
  --after hard_reset write_flash \
  --flash_mode dio \
  --flash_freq 40m \
  --flash_size detect \
  0x10000 ota_data_initial.bin \
  0x1000 bootloader/bootloader.bin \
  0x20000 at_customize.bin \
  0x21000 customized_partitions/ble_data.bin \
  0x24000 customized_partitions/server_cert.bin \
  0x26000 customized_partitions/server_key.bin \
  0x28000 customized_partitions/server_ca.bin \
  0x2a000 customized_partitions/client_cert.bin \
  0x2c000 customized_partitions/client_key.bin \
  0x2e000 customized_partitions/client_ca.bin \
  0x30000 customized_partitions/factory_param.bin \
  0xf000 phy_init_data.bin 0x100000 esp-at.bin \
  0x8000 partitions_at.bin
```

Reviewer: This is also missing the serial port, baud rate, etc., which should be added just for a working example even if it's a bit OS-specific in the case of the serial port.

Author: It is possible to use command without those parameters, esptool will search the board automagically, altough it is possible to supply port and baud-rate, I added these section in the docs.

.. note::
   Boards that already have a network interface:  Check the network
	documentation to understand how to properly configure both interfaces.
	To keep things simple, make sure WiFi is the only interface enabled in
	Networking -> Link Layer Options.  This will avoid problems running

> 	Zephyr samples.

Reviewer: Whitespace

Build and Programming
*********************

> Set ``-DSHIELD=<mikroe_wiffi_bt>`` when you invoke ``west build``.

Reviewer: ```suggestion
Set ``-DSHIELD=<mikroe_wifi_bt>`` when you invoke ``west build``.
```

Reviewer: Spelling (`wiffi`), but is the shield name `mikroe_wifi_bt` or `mikroe_wifi_bt_click`?

To build shield with specific overlay:

.. zephyr-app-commands::
   :zephyr-app: samples/net/wifi
   :board: lpcxpresso55s69_cpu0

>    :shield: mikroe_wifi_bt

Reviewer: Should this be:

```suggestion
   :shield: mikroe_wifi_bt_click
```

???

Author: It seems I left this way while I was deciding the name. :D

Alternatively, it is possible to select the port and the baud rate:

.. code-block:: console
esptool.py --chip auto --port <Selected PORT>  --baud 115200 --before default_reset --after hard_reset write_flash --flash_mode dio --flash_freq 40m --flash_size detect 0x10000 ota_data_initial.bin 0x1000 bootloader/bootloader.bin 0x20000 at_customize.bin 0x21000 customized_partitions/ble_data.bin 0x24000 customized_partitions/server_cert.bin 0x26000 customized_partitions/server_key.bin 0x28000 customized_partitions/server_ca.bin 0x2a000 customized_partitions/client_cert.bin 0x2c000 customized_partitions/client_key.bin 0x2e000 customized_partitions/client_ca.bin 0x30000 customized_partitions/factory_param.bin 0xf000 phy_init_data.bin 0x100000 esp-at.bin 0x8000 partitions_at.bin

> After flashed, it is possible to verify the module, while connnected, open your preferred terminal configured as 115200, 8, N, 1 and
> perform a board reset, you should see an initial log and it last message should be the version of the AT firmware flashed.

Reviewer 4: ```suggestion
Once flashed, it is possible to verify the module. While connected, open your preferred
terminal configured as 115200, 8, N, 1 and perform a board reset. You should see an
initial log and last message should be the version of the AT firmware flashed.
```
